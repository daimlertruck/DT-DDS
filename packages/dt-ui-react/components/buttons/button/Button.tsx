import { BaseProps } from '../../../types';
import { Spinner } from '../../spinner';

import { ButtonStyled } from './Button.styled';
import { ContainedColors, OutlinedColors, TextColors } from './constants';

export type ButtonProps = {
  onClick: () => void;
  isDisabled?: boolean;
  isLoading?: boolean;
} & BaseProps &
  (
    | { variant?: 'contained'; color?: ContainedColors }
    | { variant?: 'outlined'; color?: OutlinedColors }
    | { variant?: 'text'; color?: TextColors }
  );

export const Button = ({
  style,
  variant = 'contained',
  children,
  onClick,
  dataTestId,
  isDisabled = false,
  isLoading,
  color = 'primary',
}: ButtonProps) => {
  return (
    <ButtonStyled
      color={color}
      disabled={isDisabled}
      style={style}
      variant={variant}
      onClick={onClick}
      isLoading={isLoading}
      data-testid={`${dataTestId}-button`}
    >
      {children}
      {isLoading && (
        <Spinner
          colorScheme={variant !== 'contained' ? 'positive' : 'negative'}
          size='small'
        />
      )}
    </ButtonStyled>
  );
};
