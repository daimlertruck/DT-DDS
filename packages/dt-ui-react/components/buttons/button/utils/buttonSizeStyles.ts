import { CustomTheme as Theme } from '../../../../types';
import { ButtonSizes } from '../constants/types';

export const buttonSizeStyles = (size: ButtonSizes, theme: Theme) =>
  ({
    large: `
      ${theme.fontStyles.button1}
      padding: ${theme.spacing['2xs']} ${theme.spacing.s};
    `,
    medium: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['3xs']} ${theme.spacing.xs};
    `,
    small: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['4xs']} ${theme.spacing['2xs']};
    `,
  }[size]);
