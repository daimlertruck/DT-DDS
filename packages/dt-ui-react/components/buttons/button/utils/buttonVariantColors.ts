import { CustomTheme as Theme } from '../../../../types';
import {
  ButtonVariant,
  ContainedColors,
  OutlinedColors,
  TextColors,
} from '../constants/types';

interface ColorKeys {
  color: string;
  bgColor: string;
  hoverBgColor?: string;
  hoverColor?: string;
}

export const buttonVariantColors = (
  color: ContainedColors | OutlinedColors,
  theme: Theme
): Record<ButtonVariant, ColorKeys> => ({
  contained: {
    color: theme.colors.alpha.white,
    bgColor: theme.palette.alpha[color],
    hoverBgColor: theme.palette.alpha[`${color}Dark_300` as ContainedColors],
  },
  outlined: {
    color:
      color === 'neutral'
        ? theme.colors.alpha.gray_200
        : theme.palette.alpha[color],
    bgColor: theme.colors.alpha.white,
    hoverBgColor: theme.palette.alpha[`${color}Light_300` as OutlinedColors],
  },
  text: {
    color: theme.palette.alpha[color],
    bgColor: 'transparent',
    hoverColor: theme.palette.alpha[`${color as TextColors}Dark_300`],
  },
});
