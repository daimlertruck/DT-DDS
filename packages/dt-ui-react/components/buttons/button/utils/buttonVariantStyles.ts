import { CustomTheme as Theme } from '../../../../types';
import { ButtonVariant, OutlinedColors, SolidColors } from '../constants/types';

export const buttonVariantStyles = (
  variant: ButtonVariant,
  theme: Theme,
  color: SolidColors | OutlinedColors
) =>
  ({
    solid: `
      text-transform: uppercase;
      border: ${theme.spacing.none};

      &:focus {
        box-shadow: 0px 2px 1px 0px #1A1A1A0D, 0px 0px 0px 4px ${theme.colors.blue_10};
        background-color: ${theme.palette[color]};
      }

      &:disabled {
        opacity: 38%;
      }
  `,
    outlined: `
      border: 1px solid;
      text-transform: uppercase;
        ${
          color === 'neutral'
            ? theme.palette.neutralDark_100
            : theme.palette[color]
        };

      &:disabled {
        color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        background-color: ${theme.colors.white};
        border-color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        cursor: default;
      }
    `,
    text: `
      border: none;

      &:disabled {
        color: ${theme.palette.neutralLight_100};
      }
    `,
  }[variant]);
