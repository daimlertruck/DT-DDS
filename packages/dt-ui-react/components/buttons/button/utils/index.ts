export * from './buttonVariantColors';
export * from './buttonVariantStyles';
