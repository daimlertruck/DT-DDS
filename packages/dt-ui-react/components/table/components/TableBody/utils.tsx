import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { TableDataCellHeader } from '../TableDataCell/TableDataCell';

export const addHeadersToDataCells = (
  dataCells: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(dataCells, (dataCell, index) => {
    if (isValidElement(dataCell)) {
      const header = headers[index];
      return cloneElement(
        dataCell,
        {},
        <>
          <TableDataCellHeader>{header}</TableDataCellHeader>
          {dataCell.props.children}
        </>
      );
    }
    return dataCell;
  });
};

export const addHeadersToDataCellsInRows = (
  rows: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(rows, (row) => {
    if (isValidElement(row)) {
      return cloneElement(
        row,
        {},
        addHeadersToDataCells(row.props.children, headers)
      );
    }
    return row;
  });
};
