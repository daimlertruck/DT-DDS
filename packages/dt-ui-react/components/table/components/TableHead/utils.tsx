import { Children, ReactNode, isValidElement } from 'react';

const extractColumnHeaders = (columnHeaders: ReactNode): ReactNode[] => {
  return Children.toArray(columnHeaders).reduce<ReactNode[]>(
    (headers, columnHeader) => {
      if (isValidElement(columnHeader)) {
        return [...headers, columnHeader.props.children];
      }

      return headers;
    },
    []
  );
};

export const extractTableHeaders = (theadRow: ReactNode) => {
  if (isValidElement(theadRow)) {
    return extractColumnHeaders(theadRow.props.children);
  }

  return [];
};
