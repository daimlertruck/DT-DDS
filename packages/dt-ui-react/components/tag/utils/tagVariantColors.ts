import { CustomTheme as Theme } from '../../../types';
import { TagColor, TagVariant } from '../constants';

interface TagVariantColors {
  border: string | number;
  color: string;
  borderColor?: string;
  backgroundColor: string;
  '&:hover'?: {
    borderColor?: string;
    backgroundColor: string;
    color: string;
  };
}

interface TagVariantColorsProps {
  theme: Theme;
  color: TagColor;
  variant: TagVariant;
  hasHover: boolean;
  isDisabled: boolean;
}

const getColorWithShade = ({
  theme,
  color,
  shade,
}: Pick<TagVariantColorsProps, 'color'> & {
  shade: '90' | '60' | '30' | '10';
  theme: Theme;
}) => {
  const { colors, palette } = theme;

  if (color === 'primary') {
    return {
      '90': palette.primaryDark_500,
      '60': palette.primaryDark_200,
      '30': palette.primaryLight_100,
      '10': palette.primaryLight_400,
    }[shade];
  }

  const colorWithShade: keyof Theme['colors'] = `${color}_${shade}`;

  return colors[colorWithShade];
};

export const tagVariantColors = ({
  theme,
  color,
  variant,
  hasHover,
  isDisabled,
}: TagVariantColorsProps): TagVariantColors => {
  const colorShade90 = getColorWithShade({
    theme,
    color,
    shade: '90',
  });
  const colorShade60 = getColorWithShade({
    theme,
    color,
    shade: '60',
  });
  const colorShade30 = getColorWithShade({
    theme,
    color,
    shade: '30',
  });
  const colorShade10 = getColorWithShade({
    theme,
    color,
    shade: '10',
  });

  const colorWithDisabled = isDisabled ? colorShade30 : colorShade60;

  return {
    solid: {
      border: theme.spacing.none,
      color: theme.colors.white,
      backgroundColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          backgroundColor: colorShade90,
          color: theme.colors.white,
        },
      }),
    },
    outlined: {
      border: `1px solid ${colorShade60}`,
      color: colorWithDisabled,
      backgroundColor: colorShade10,
      borderColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          borderColor: colorShade90,
          backgroundColor: colorShade10,
          color: colorShade90,
        },
      }),
    },
  }[variant];
};
