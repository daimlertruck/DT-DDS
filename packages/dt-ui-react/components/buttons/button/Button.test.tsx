import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

import { withProviders } from '../../../utils';

import { Button } from './Button';

describe('<Button /> component', () => {
  const ProvidedButton = withProviders(Button);

  it('renders button html element with the correct title', () => {
    const { container } = render(
      <ProvidedButton onClick={() => null} dataTestId='primary'>
        Primary
      </ProvidedButton>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a disabled button html element with the correct title', () => {
    render(
      <ProvidedButton
        isDisabled={true}
        variant='outlined'
        color='neutral'
        onClick={() => null}
        dataTestId='secondary'
      >
        Secondary
      </ProvidedButton>
    );

    const secondaryButton = screen.getByTestId('secondary-button');
    expect(secondaryButton).toHaveTextContent('Secondary');
    expect(secondaryButton).toBeDisabled();
  });

  it('triggers the button action correctly', () => {
    const callbackFn = jest.fn();

    render(
      <ProvidedButton
        variant='contained'
        color='error'
        onClick={callbackFn}
        dataTestId='error'
      >
        Error
      </ProvidedButton>
    );

    const button = screen.getByTestId('error-button');
    fireEvent.click(button);

    expect(callbackFn).toBeCalledTimes(1);
  });

  describe('when loading is true', () => {
    it('renders a spinner light theme', () => {
      const { container } = render(
        <ProvidedButton isLoading={true} onClick={() => null}>
          Button loading
        </ProvidedButton>
      );

      expect(container).toMatchSnapshot();
    });

    it('renders a spinner dark theme when variant is outlined', () => {
      const { container } = render(
        <ProvidedButton
          isLoading={true}
          onClick={() => null}
          variant='outlined'
        >
          Button loading
        </ProvidedButton>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
