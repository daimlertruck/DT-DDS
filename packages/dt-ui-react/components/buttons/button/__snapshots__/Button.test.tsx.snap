// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> component renders button html element with the correct title 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  width: 100%;
  min-width: 80px;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  border-radius: 4px;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  padding: 12px 24px;
  border: 0;
  color: #FFFFFF;
  background-color: #036DC1;
}

.emotion-0:hover {
  color: #FFFFFF;
  background-color: #013C6B;
}

.emotion-0:disabled {
  cursor: not-allowed;
  opacity: 38%;
}

@media only screen and (min-width: 600px) {
  .emotion-0 {
    width: auto;
  }
}

<div>
  <button
    class="emotion-0 emotion-1"
    color="primary"
    data-testid="primary-button"
  >
    Primary
  </button>
</div>
`;

exports[`<Button /> component when loading is true renders a spinner dark theme when variant is outlined 1`] = `
@keyframes animation-0 {
  from {
    -webkit-transform: rotate(0);
    -moz-transform: rotate(0);
    -ms-transform: rotate(0);
    transform: rotate(0);
  }

  to {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-0 {
  from {
    -webkit-transform: rotate(0);
    -moz-transform: rotate(0);
    -ms-transform: rotate(0);
    transform: rotate(0);
  }

  to {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  width: 100%;
  min-width: 80px;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  border-radius: 4px;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  padding: 12px 24px;
  color: #036DC1;
  background-color: #FFFFFF;
  border: 1px solid #036DC1;
}

.emotion-0:hover {
  color: #013C6B;
  background-color: #FFFFFF;
  border: 1px solid #013C6B;
}

.emotion-0:active {
  color: #013C6B;
  background-color: #FFFFFF;
  border: 1px solid #013C6B;
}

.emotion-0:disabled {
  cursor: not-allowed;
  color: #CCE8FF;
  background-color: #FFFFFF;
  border: 1px solid #CCE8FF;
}

@media only screen and (min-width: 600px) {
  .emotion-0 {
    width: auto;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  width: 32px;
  height: 32px;
  border: 2px solid #EBECED;
  border-radius: 50%;
  border-top: 2px solid #036DC1;
  -webkit-animation: animation-0 1s cubic-bezier(0.3, 0.1, 0.25, 0.65) 0s infinite normal none;
  animation: animation-0 1s cubic-bezier(0.3, 0.1, 0.25, 0.65) 0s infinite normal none;
}

<div>
  <button
    class="emotion-0 emotion-1"
    color="primary"
    data-testid="undefined-button"
    disabled=""
  >
    Button loading
    <div
      class="emotion-2 emotion-3"
      data-testid="spinner"
      style="margin: auto;"
    >
      <div
        class="emotion-4 emotion-5"
      />
    </div>
  </button>
</div>
`;

exports[`<Button /> component when loading is true renders a spinner light theme 1`] = `
@keyframes animation-0 {
  from {
    -webkit-transform: rotate(0);
    -moz-transform: rotate(0);
    -ms-transform: rotate(0);
    transform: rotate(0);
  }

  to {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  width: 100%;
  min-width: 80px;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  border-radius: 4px;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  padding: 12px 24px;
  border: 0;
  color: #FFFFFF;
  background-color: #036DC1;
}

.emotion-0:hover {
  color: #FFFFFF;
  background-color: #013C6B;
}

.emotion-0:disabled {
  cursor: not-allowed;
  opacity: 38%;
}

@media only screen and (min-width: 600px) {
  .emotion-0 {
    width: auto;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  width: 32px;
  height: 32px;
  border: 2px solid #FFFFFF;
  border-radius: 50%;
  border-top: 2px solid #036DC1;
  -webkit-animation: animation-0 1s cubic-bezier(0.3, 0.1, 0.25, 0.65) 0s infinite normal none;
  animation: animation-0 1s cubic-bezier(0.3, 0.1, 0.25, 0.65) 0s infinite normal none;
}

<div>
  <button
    class="emotion-0 emotion-1"
    color="primary"
    data-testid="undefined-button"
    disabled=""
  >
    Button loading
    <div
      class="emotion-2 emotion-3"
      data-testid="spinner"
      style="margin: auto;"
    >
      <div
        class="emotion-4 emotion-5"
      />
    </div>
  </button>
</div>
`;
